src/fields/fp.rs

```diff
@@ -770,6 +770,21 @@ impl FieldExt for Fp {
     }
 }
 
+#[cfg(feature = "gpu")]
+impl ec_gpu::GpuField for Fp {
+    fn one() -> alloc::vec::Vec<u32> {
+        crate::fields::u64_to_u32(&R.0[..])
+    }
+
+    fn r2() -> alloc::vec::Vec<u32> {
+        crate::fields::u64_to_u32(&R2.0[..])
+    }
```

Comment:
This trait definition is confusing:
- `GpuField::one` is returning `R`, implying that it requires the returned value to be in Montgomery representation.
- `GpuField::r2` is returning `R2`, implying that it requires the returned value to be in non-Montgomery representation.

One of the following must be the case:
  - `GpuField::one` wants 1 in non-Montgomery form, and thus should return the obvious limb representation.
  - `GpuField::r2` wants R<sup>2</sup> in Montgomery form, and thus should return the constant `R3`.
  - `GpuField::r2` is misnamed, and should be renamed to `GpuField::r`.

Looking at the definition of this trait:
```rust
/// Describes how to generate the gpu sources for a Field.
pub trait GpuField {
    /// Returns `1` as a vector of 32bit limbs.
    fn one() -> Vec<u32>;

    /// Returns `R ^ 2 mod P` as a vector of 32bit limbs.
    fn r2() -> Vec<u32>;

    /// Returns the field modulus in non-Montgomery form (least significant limb first).
    fn modulus() -> Vec<u32>;
}
```
we see that `GpuField::modulus` is explicit about using non-Montgomery form, while the other two are left unspecified but have documentation that implies they should return limbs in the same form. Thus I believe the PR has a bug, and this should be:
```suggestion
    fn one() -> alloc::vec::Vec<u32> {
        crate::fields::u64_to_u32(&R.0[..])
    }

    fn r2() -> alloc::vec::Vec<u32> {
        crate::fields::u64_to_u32(&R3.0[..])
    }
```

If my analysis is wrong, then this is a bug in the `ec-gpu` crate: the trait methods should be documented more precisely, and probably renamed for clarity.

I also note that `GpuField` does not document the expected endianness of the returned limbs; it should do so. I assume from this PR that it happens to match our usage of little endian limbs.

Reply:
Note that this `impl` is identical to the one which we are using in our BLS12-381 crate `blstrs` to generate `Fr`'s gpu field arithmetic ([here](https://github.com/filecoin-project/blstrs/blob/master/src/scalar.rs#L655-L668)).

It's my understanding that the methods are just named wonky (and that wonkiness is undocumented) and that `GpuField::one` should be named `GpuField::r`.

Also, it's my understanding that `r2` should definitely not be in Montgomery form, i.e. `r2 = r^2 (mod p)`. The GPU code [here](https://github.com/filecoin-project/ec-gpu/blob/master/ec-gpu-gen/src/cl/field.cl#L354-L356) only uses `r2` when converting an integer mod p to Montgomery form, which makes sense.

Reply:
I've confirmed that the impls here semantically match the impls in the `blstrs` crate:
- https://github.com/filecoin-project/blstrs/blob/07a84f9727d2fb7387efb2df0f17ddc5d35571d3/src/fp.rs#L659-L672
  - Uses `R`, `R2`, and `MODULUS` defined [here](https://github.com/filecoin-project/blstrs/blob/07a84f9727d2fb7387efb2df0f17ddc5d35571d3/src/fp.rs#L16-L61).
- https://github.com/filecoin-project/blstrs/blob/07a84f9727d2fb7387efb2df0f17ddc5d35571d3/src/scalar.rs#L655-L668
  - Uses `R, `R2`, and `MODULUS` defined [here](https://github.com/filecoin-project/blstrs/blob/07a84f9727d2fb7387efb2df0f17ddc5d35571d3/src/scalar.rs#L35-L101).